credentials/tls/certprovider/store.go

```diff
@@ -0,0 +1,126 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package certprovider
+
+import (
+	"sync"
+
+	"google.golang.org/grpc/grpclog"
+)
+
+// store is the global singleton certificate provider store.
+var store *Store
+
+func init() {
+	store = &Store{
+		providers: make(map[storeKey]*wrappedProvider),
+	}
+}
+
+// Key contains data which uniquely identifies a provider instance.
+type Key struct {
+	// Name is the registered name of the provider.
+	Name string
+	// Config is the configuration used by the provider instance.
+	Config StableConfig
+}
+
+// storeKey acts the key to the map of providers maintained by the store. Go
+// maps need to be indexed by comparable types, so the above Key struct cannot
+// be used since it contains an interface.
+type storeKey struct {
+	// name of the certificate provider.
+	name string
+	// configuration of the certificate provider in string form.
+	config string
+}
+
+// wrappedProvider wraps a provider instance with a reference count.
+type wrappedProvider struct {
+	Provider
+	refCount int
+
+	// A reference to the key and store are also kept here to override the
+	// Close method on the provider.
+	storeKey storeKey
+	store    *Store
+}
+
+// Store is a collection of provider instances, safe for concurrent access.
+type Store struct {
+	mu        sync.Mutex
+	providers map[storeKey]*wrappedProvider
+}
+
+// GetStore returns a global singleton store of provider instances.
+func GetStore() *Store {
```

Comment:
`GlobalStore`?  `DefaultStore`?

Reply:
I initially had it as `DefaultStore`. But we decided that since `Store` does not have an exported `New` method, there is no way for callers outside of the package to create a `Store` other than the one returned from here. And if other packages want to fake out the store in their tests, we decided that they can define an interface and provide a fake implementation themselves.